pennylane/numpy/tensor.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module provides the PennyLane :class:`~.tensor` class.
+"""
+import numpy as onp
+
+from autograd import numpy as _np
+
+from autograd.tracer import Box
+from autograd.numpy.numpy_boxes import ArrayBox
+from autograd.numpy.numpy_vspaces import ComplexArrayVSpace, ArrayVSpace
+from autograd.core import VSpace
+
+
+__doc__ = "NumPy with automatic differentiation support, provided by Autograd and PennyLane."
+
+
+class tensor(_np.ndarray):
+    """Constructs a PennyLane tensor for use with Autograd QNodes.
+
+    The ``tensor`` class is a subclass of ``numpy.ndarray``,
+    providing the same multidimensional, homogeneous data-structure
+    of fixed-size items, with an additional flag to indicate to PennyLane
+    whether the contained data is differentiable or not.
+
+    .. warning::
+
+        PennyLane ``tensor`` objects are only used as part of the Autograd QNode
+        interface. If using another machine learning library such as PyTorch or
+        TensorFlow, use their built-in ``tf.Variable`` and ``torch.tensor`` classes
+        instead.
+
+    .. warning::
+
+        Tensors should be constructed using standard array construction functions
+        provided as part of PennyLane's NumPy implementation, including
+        ``np.array``, ``np.zeros`` or ``np.empty``.
+
+        The parameters given here refer to a low-level class
+        for instantiating an tensors.
+
+    Args:
+        input_array (array_like): Any data structure in any form that can be converted to
+            an array. This includes lists, lists of tuples, tuples, tuples of tuples,
+            tuples of lists and ndarrays.
+        requires_grad (bool): whether the tensor supports differentiation
+
+    **Example**
+
+    The trainability of a tensor can be set on construction via the
+    ``requires_grad`` keyword argument,
+
+    >>> from pennylane import numpy as np

```

Comment:
I can't remember, do we recommend using `code-block` in these kind of docstrings?

Reply:
Nope, we recommend `pycon` without `.. code-block::` directives. This is the [guideline followed by NumPy/SciPy](https://numpy.org/devdocs/docs/howto_document.html#docstring-standard) (see point (15)), and is compatible with `doctest` if we decide to test our documentation examples in the future.